Fix Second-Chance Algorithm issue

hello2mao opened the pull request:
#### Type of change

- Bug fix

#### Description

New added item's referenced bit should be set to false when cache is not full.

#### Reproduce:
```golang
func TestSecondChanceCache(t *testing.T) {
	cache := newSecondChanceCache(2)
	assert.NotNil(t, cache)

	cache.add("a", "xyz")

	cache.add("b", "123")
	// get b, b referenced bit is set to true
	obj, ok := cache.get("b")
	assert.True(t, ok)
	assert.Equal(t, "123", obj.(string))

	// add c. victim scan: delete a and set b as the next candidate of a victim
	cache.add("c", "777")

	// check a is deleted
	_, ok = cache.get("a")
	assert.False(t, ok)

	// add d. victim scan: b referenced bit is set to false and delete c
	cache.add("d", "555")

	// check c is deleted
	_, ok = cache.get("c")
	assert.False(t, ok)

	// check b and d
	obj, ok = cache.get("b")
	assert.True(t, ok)
	assert.Equal(t, "123", obj.(string))
	obj, ok = cache.get("d")
	assert.True(t, ok)
	assert.Equal(t, "555", obj.(string))
}
```

**Result**:

test failed when check c is deleted.

**Expect**:

test success.

**Explain**:

- When add c, delete a and set b as the next candidate of a victim.
- When add d, b referenced bit changes from true to false and c should be deleted.
- But the result is: b is deleted instead of c.

#### Related issues

NA

#### Release Note

NA

srderson commented:
LGTM. Thanks

hello2mao commented:
@yoheiueda
Could you help me review this pr?

hello2mao commented:
/ci-run

sykesm commented on msp/cache/second_chance_test.go:
```diff
@@ -14,7 +14,7 @@ import (
 )
 
-func TestSecondChanceCache(t *testing.T) {
+func TestSecondChanceCacheCase1(t *testing.T) {
```
Some questions based on very quick glance at the change:

What do `Case1` and `Case2` mean?
Can we use more descriptive names for the scenarios?
Are the scenarios different enough as to require separate tests?

Looks to me like the new test can simply replace the old one.

hello2mao replied:
OK. I keep the new test.

sykesm commented on msp/cache/second_chance.go:
```diff
@@ -82,7 +82,7 @@ func (cache *secondChanceCache) add(key string, value interface{}) {
 	item.key = key
 	item.value = value
-	atomic.StoreInt32(&item.referenced, 1)
+	atomic.StoreInt32(&item.referenced, 0)
```
fwiw, should be no need for the atomic store as the zero value is already 0.

hello2mao replied:
Thanks. I just remove this line.

sykesm reviewed (commented):
Given the classic cache implementation does not set reference on initial population, the change looks okay; it certainly doesn't break anything.

Biggest questions are around test names and why we need two instead of one.

Thanks.

hello2mao committed:
Fix Second-Chance Algorithm issue.

New added item's referenced bit should be set to false
when cache is not full.

Signed-off-by: Hongbin Mao <hello2mao@gmail.com>
